### PR TITLE
Retry to get session when session is invalid in `SessionPool`

### PR DESCRIPTION
**Pull request by Aiee**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

#### Issue(s) number: 
Close https://github.com/vesoft-inc/nebula-go/issues/248

### CODE REVIEW

**Comment by wey-gu** [comment]
we may need such retry mechanism in all clients :D

File: session_pool.go [hunk 1]
```diff
@@ -384,6 +394,42 @@ func (pool *SessionPool) getIdleSession() (*Session, error) {
...
+		}
+
+		pingErr := newSession.Ping()
+		if pingErr != nil {
+			pool.log.Error("failed to ping the session, error: " + pingErr.Error())
+			continue
+		}
+		pool.log.Info("retry to get sessions successfully")
+		pool.addSessionToList(&pool.activeSessions, newSession)
+		pool.removeSessionFromList(&pool.activeSessions, session)
```
[the remark this answers is not in the document]
Comment (Aiee) [reply session_pool.go]: Fixed.

File: session_pool.go [hunk 2]
```diff
@@ -384,6 +394,42 @@ func (pool *SessionPool) getIdleSession() (*Session, error) {
...
+		return f(newSession)
+	}
+	pool.log.Error(fmt.Sprintf("failed to get session after " + strconv.Itoa(retry) + " retries"))
+	return nil, err
```
[the remark this answers is not in the document]
Comment (Aiee) [reply session_pool.go]: Fixed.

**Review by HarrisChu — APPROVED** [review]

**Comment by wenhaocs** [comment]
Can you merge this PR to master? We are still working on the http support for `beta-sc`, which may or may not be used. In addition, I think this feature is useful for all. Thanks!

File: session_pool.go [hunk 3]
```diff
@@ -384,6 +394,50 @@ func (pool *SessionPool) getIdleSession() (*Session, error) {
 		" session pool and the total session count has reached the limit")
 }
 
+// retryGetSession tries to create a new session when the current session is invalid.
+func (pool *SessionPool) executeWithRetry(
+	session *Session,
+	f func(*Session) (*graph.ExecutionResponse, error),
+	retry int) (*graph.ExecutionResponse, error) {
+	pool.rwLock.Lock()
+	defer pool.rwLock.Unlock()
```
Comment (veezhang) [inline session_pool.go:403]: Should it be possible to reduce the granularity of locks?
Only need to lock *list.List, not during session creation, execution, etc.

Comment (Aiee) [reply session_pool.go:403]: Yeah, we should recheck the lock usage later.

**Pull request merged**